internal/security/secretstore/tokenmaintenance.go

```diff
@@ -0,0 +1,189 @@
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+//
+// SPDX-License-Identifier: Apache-2.0'
+//
+
+package secretstore
+
+/*
+
+High level token maintenance flow is as follows:
+
+1. Recover a root token from keyshares
+2. Revoke all previous non-root tokens
+3. Create new admin token from that root token
+4. Create per-service tokens using admin token
+5. Upload other bootstrapping access with root token (e.g. kong cert, db password)
+6. Revoke all other root tokens (dangerous: breaks backward-compatibility)
+   6a. Go back and retroactively remove root token from resp-init.json
+7. Revoke self
+
+End state is that only admin token and per-service tokens remain.
+For Fuji, the root token in resp-init.json will also remain.
+For Fuji.DOT/Geneva, all root tokens will be revoked.
+
+*/
+
+import (
+	"fmt"
+
+	"github.com/edgexfoundry/edgex-go/internal/security/secretstoreclient"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+)
+
+type RevokeFunc func()
+
+type TokenMaintenance struct {
+	logging      logger.LoggingClient
+	secretClient secretstoreclient.SecretStoreClient
+}
+
+// NewTokenMaintenance creates a new TokenProvider
+func NewTokenMaintenance(logging logger.LoggingClient, secretClient secretstoreclient.SecretStoreClient) *TokenMaintenance {
+	return &TokenMaintenance{
+		logging:      logging,
+		secretClient: secretClient,
+	}
+}
+
+// CreateTokenIssuingToken creates an admin token that
+// allows the holder to create per-service tokens an policies.
+// Requires a root token, returns a function that,
+// if called, with revoke the token
+func (tm *TokenMaintenance) CreateTokenIssuingToken(rootToken string) (map[string]interface{}, RevokeFunc, error) {
+
+	_, err := tm.secretClient.InstallPolicy(rootToken, TokenCreatorPolicyName, TokenCreatorPolicy)
+	if err != nil {
+		tm.logging.Error("failed installation of token-issuing-token policy")
+		return nil, nil, err
+	}
+
+	createTokenParameters := make(map[string]interface{})
+	createTokenParameters["display_name"] = TokenCreatorPolicyName
+	createTokenParameters["no_parent"] = true
+	createTokenParameters["period"] = "1h"
+	createTokenParameters["policies"] = []string{TokenCreatorPolicyName}
+	createTokenParameters["ttl"] = "1h"
+	createTokenResponse := make(map[string]interface{})
+	_, err = tm.secretClient.CreateToken(rootToken, createTokenParameters, &createTokenResponse)
+	if err != nil {
+		tm.logging.Error(fmt.Sprintf("failed creation of token-issuing-token: %s", err.Error()))
+		return nil, nil, err
+	}
+
+	newToken := createTokenResponse["auth"].(map[string]interface{})["client_token"].(string)
+	revokeFunc := func() {
+		tm.logging.Info("revoking token-issuing-token")
+		if _, err2 := tm.secretClient.RevokeSelf(newToken); err2 != nil {
```

Comment:
Not sure what you mean.  We will parse the response for the token, and create a closure with revokeFunc to revoke it.  Ultimately, the json response will be serialized to disk and passed to the token provider via a file.  I explicitly extract newToken before declaring the function so that the caller can't alter the response object and affect it.

Reply:
Ok, I noticed that `newToken` is the variable created outside this `revokeFunc` and that is got "closured" into that func.  The `revokeFunc` is not called at this time and it is called later hence the binding later.